File: library/src/main/java/cn/wandersnail/http/download/ProgressInterceptor.java
package cn.wandersnail.http.download;


import androidx.annotation.NonNull;

import java.io.IOException;

import cn.wandersnail.http.callback.ProgressListener;
import okhttp3.Interceptor;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * 拦截器，加入进度监听
 * <p>
 * date: 2019/8/23 16:31
 * author: zengfansheng
 */
class ProgressInterceptor implements Interceptor {
    private final ProgressListener listener;

    ProgressInterceptor(ProgressListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public Response intercept(Chain chain) throws IOException {
        Response originalResponse = chain.proceed(chain.request());
        ResponseBody body = originalResponse.body();
        ResponseBody progressBody = body == null ? null : new ProgressResponseBody(body, listener);
        return originalResponse.newBuilder().body(progressBody).build();
    }
}

File: library/src/main/java/com/snail/network/download/ProgressInterceptor.kt
package com.snail.network.download

import com.snail.network.callback.ProgressListener
import okhttp3.Interceptor
import okhttp3.Response

import java.io.IOException

/**
 * 时间: 2017/7/6 14:47
 * 作者: zengfansheng
 * 邮箱: 43068145@qq.com
 * 功能: 拦截器，加入进度监听
 * 
 * @param listener 进度
 */

internal class ProgressInterceptor(private val listener: ProgressListener) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val originalResponse = chain.proceed(chain.request())
        val body = originalResponse.body()
        return originalResponse.newBuilder().body(if (body == null) null else ProgressResponseBody(body, listener)).build()
    }
}
